### PR TITLE
Fix 59191 - ICE when macro replaces crate root with non-module item

### LINKED ISSUE #59191: thread 'rustc' panicked at 'internal error: entered unreachable code', src/libsyntax/ext/expand.rs:289:18
This is nonsense code, but it probably shouldn't ICE.

```
$ cargo run --example example
   Compiling no_main v0.1.0 (/home/xfix/kod/no_main)
error[E0658]: non-builtin inner attributes are unstable (see issue #54726)
 --> examples/example.rs:1:1
  |
1 | #![no_main::no_main]
  | ^^^^^^^^^^^^^^^^^^^^

error[E0658]: custom attributes cannot be applied to modules (see issue #54727)
 --> examples/example.rs:1:1
  |
1 | #![no_main::no_main]
  | ^^^^^^^^^^^^^^^^^^^^

error[E0468]: an `extern crate` loading macros must be at the crate root

thread 'rustc' panicked at 'internal error: entered unreachable code', src/libsyntax/ext/expand.rs:289:18
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to 3 previous errors

Some errors occurred: E0468, E0658.
For more information about an error, try `rustc --explain E0468`.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.33.0 (2aa4c46cf 2019-02-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `no_main`.

To learn more, run the command again with --verbose.
```

# Cargo.toml

```toml
[package]
name = "no_main"
version = "0.1.0"
edition = "2018"

[lib]
proc-macro = true
```

# src/lib.rs

```rust
extern crate proc_macro;

use proc_macro::{Delimiter, Group, Ident, Span, TokenStream, TokenTree};

#[proc_macro_attribute]
pub fn no_main(_attrs: TokenStream, input: TokenStream) -> TokenStream {
    let mut ts = TokenStream::new();
    ts.extend(vec![
        TokenTree::Ident(Ident::new("fn", Span::call_site())),
        TokenTree::Ident(Ident::new("main", Span::call_site())),
        TokenTree::Group(Group::new(Delimiter::Parenthesis, TokenStream::new())),
        TokenTree::Group(Group::new(Delimiter::Brace, input)),
    ]);
    ts
}
```

# examples/example.rs

```rust
#![no_main::no_main]
```

### PR DESCRIPTION
**Pull request by daboross**
Hi,

This should fix #59191! My friend and I are working on learning the rustc codebase through contributions, so please feel free to mention anything amiss or that could be done better.

The code adds an explicit case for when a macro applied to the crate root (via an inner attribute) replaces it with something nonsensical, like a function. The crate root must be a module, and the error message reflects this.

---

I should note that there are a few other weird edge cases here, like if they do output a module, it succeeds but uses that module's name as a prefix for all names in the crate. I'm assuming that's an issue for stabilizing #54726, though.

### CODE REVIEW

File: src/librustc_expand/expand.rs [hunk 1]
```diff
@@ -363,7 +363,15 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                 krate.attrs = vec![];
                 krate.module = ast::Mod { inner: orig_mod_span, items: vec![], inline: true };
             }
-            _ => unreachable!(),
+            Some(ast::Item { span, kind, .. }) => {
+                self.cx.span_fatal(
```
Comment (Centril) [inline src/librustc_expand/expand.rs]: Seems like a fatal error would not be necessary here? Could do what the `None =>` branch above does and then use `cx.struct_span_err(...).emit()`.

Comment (daboross) [reply src/librustc_expand/expand.rs]: Oh true! We'd read through it, but somehow hadn't absorbed the fact that return is erroring out without a fatal error.

Thanks!

File: src/librustc_expand/expand.rs [hunk 2]
```diff
@@ -363,7 +363,17 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
                 krate.module = ast::Mod { inner: orig_mod_span, items: vec![], inline: true };
             }
-            _ => unreachable!(),
+            Some(ast::Item { span, kind, .. }) => {
+                krate.attrs = vec![];
+                krate.module = ast::Mod { inner: orig_mod_span, items: vec![], inline: true };
+                self.cx.span_err(
+                    span,
+                    &format!(
+                        "expected crate top-level item to be a module after macro expansion, found a {}",
```
Comment (petrochenkov) [inline src/librustc_expand/expand.rs]: Crate-level attribute macros aren't really expected to produce a module, they are just not supported at all and may result in nonsense and paradoxes.
This is an ok fix and it removes the ICE, but this case should ideally report the same error as this example:
```rust
enum E {
    #[test] // error: expected an inert attribute, found an attribute macro
    V
}

fn main() {}
```

Comment (petrochenkov) [reply src/librustc_expand/expand.rs]: It's not urgent though, and I still hope to rewrite the macro invocation collector anyway.

**Pull request merged**